python/paddle/fluid/tests/unittests/test_optimizer_in_control_flow.py

```diff
@@ -0,0 +1,195 @@
+# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import numpy as np
+import unittest
+
+import paddle.fluid as fluid
+import paddle.fluid.layers as layers
+import paddle.fluid.optimizer as optimizer
+
+BATCH_SIZE = 1
+INPUT_SIZE = 784
+
+CLASS_NUM = 10
+FC_SIZE = 40
+EPOCH_NUM = 20
+Switch_ID = 3
+
+
+def random_input(seed,
+                 image_shape=[BATCH_SIZE, INPUT_SIZE],
+                 label_shape=[BATCH_SIZE, 1]):
+    np.random.seed(seed)
+    image_np = np.random.random(size=image_shape).astype('float32')
+    np.random.seed(seed)
+    label_np = np.random.random_integers(
+        low=0, high=CLASS_NUM - 1, size=label_shape).astype('int64')
+    return image_np, label_np
+
+
+def random_param(size, seed=100):
+    np.random.seed(seed)
+    np_param = np.random.random(size=size).astype('float32')
+    return np_param
+
+
+def static():
+    def simple_fc_net(image):
+        hidden = layers.fc(
+            image,
+            size=FC_SIZE,
+            act='relu',
+            param_attr=fluid.ParamAttr(initializer=fluid.initializer.Constant(
+                value=0.99)),
+            bias_attr=fluid.ParamAttr(initializer=fluid.initializer.Constant(
+                value=0.5)),
+            name="hidden")
+
+        prediction = layers.fc(
+            hidden,
+            size=CLASS_NUM,
+            act='softmax',
+            param_attr=fluid.ParamAttr(initializer=fluid.initializer.Constant(
+                value=1.2)),
+            bias_attr=fluid.ParamAttr(initializer=fluid.initializer.Constant(
+                value=0.8)),
+            name="prediction")
+        return hidden, prediction
+
+    image = fluid.data(
+        name='image', shape=[BATCH_SIZE, INPUT_SIZE], dtype='float32')
+    label = fluid.data(name='label', shape=[BATCH_SIZE, 1], dtype='int64')
+    switch_id = fluid.data(name='switch_id', shape=[1], dtype='int32')
+
+    id = layers.fill_constant(shape=[1], dtype='int32', value=Switch_ID)
+    hidden, prediction = simple_fc_net(image)
+
+    adam = optimizer.Adam(learning_rate=0.001)
```

Comment:
Nothing to change, but just a question, can we put the "adam = optimizer.Adam(learning_rate=0.001)" in a branch such as "fn_1" ? Maybe we can add a test case to cover that.

Reply:
yes. I will add a test case